db/db_impl/db_impl_write.cc

```diff
@@ -2019,34 +2043,47 @@ size_t DBImpl::GetWalPreallocateBlockSize(uint64_t write_buffer_size) const {
 // can call if they wish
 Status DB::Put(const WriteOptions& opt, ColumnFamilyHandle* column_family,
                const Slice& key, const Slice& value) {
-  if (nullptr == opt.timestamp) {
-    // Pre-allocate size of write batch conservatively.
-    // 8 bytes are taken by header, 4 bytes for count, 1 byte for type,
-    // and we allocate 11 extra bytes for key length, as well as value length.
-    WriteBatch batch(key.size() + value.size() + 24);
-    Status s = batch.Put(column_family, key, value);
-    if (!s.ok()) {
-      return s;
-    }
-    return Write(opt, &batch);
-  }
-  const Slice* ts = opt.timestamp;
-  assert(nullptr != ts);
-  size_t ts_sz = ts->size();
-  assert(column_family->GetComparator());
-  assert(ts_sz == column_family->GetComparator()->timestamp_size());
-  WriteBatch batch;
-  Status s;
-  if (key.data() + key.size() == ts->data()) {
-    Slice key_with_ts = Slice(key.data(), key.size() + ts_sz);
-    s = batch.Put(column_family, key_with_ts, value);
-  } else {
-    std::array<Slice, 2> key_with_ts_slices{{key, *ts}};
-    SliceParts key_with_ts(key_with_ts_slices.data(), 2);
-    std::array<Slice, 1> value_slices{{value}};
-    SliceParts values(value_slices.data(), 1);
-    s = batch.Put(column_family, key_with_ts, values);
+  column_family = column_family ? column_family : DefaultColumnFamily();
```

Comment:
I think application can still call `DB::Put(write_opts, nullptr, key, value)`, as it's a public API?

Reply:
Right, thanks! One related question: it seems to me that the timestamp-less APIs silently translate a null `ColumnFamilyHandle` to the default CF, while the timestamp-aware ones are more strict and reject calls where the `cfh` is null. Would it make sense to bring these into sync?

Reply:
> the timestamp-less APIs silently translate a null ColumnFamilyHandle to the default CF

I believe (I can be wrong) this is for backward compatibility: when column family support was added, the these APIs and those in write_batch.h were already there. If I were to design the APIs with the concept of column family from the very beginning, I would have enforced this on all APIs.

That said, when we introduce new APIs, it may be a good idea to be strict.